Frameworks/CoreGraphics/CGGradient.mm

```diff
@@ -20,85 +20,81 @@
 
-    inline void SetColorSpace(CGColorSpaceRef colorSpace) {
-        _colorSpace = colorSpace;
+    __CGGradient(CGColorSpaceRef colorSpace, CFArrayRef components, const CGFloat* locations)
+        : _colorSpace(colorSpace),
+          _colorComponents(CopyColorComponents(colorSpace, components)),
+          _stopLocations(CopyLocations(locations, CFArrayGetCount(components))) {
     }
 
-    inline CGColorSpaceRef GetColorSpace() {
+    inline CGColorSpaceRef GetColorSpace() const {
         return _colorSpace;
     }
 
-    inline std::vector<CGFloat> GetColorComponents() {
+    inline std::vector<CGFloat> GetColorComponents() const {
         return _colorComponents;
     }
 
-    inline std::vector<CGFloat> GetStopLocations() {
+    inline std::vector<CGFloat> GetStopLocations() const {
         return _stopLocations;
     }
 
-    void InitWithColorComponents(const CGFloat* components, const CGFloat* locations, size_t count);
-    void InitWithColors(CFArrayRef components, const CGFloat* locations);
-
 private:
     woc::StrongCF<CGColorSpaceRef> _colorSpace;
     std::vector<CGFloat> _colorComponents;
     std::vector<CGFloat> _stopLocations;
 
-    void SetLocations(const CGFloat* locations, size_t count) {
-        RETURN_IF(count == 0);
-
-        _stopLocations.reserve(count);
-
+    std::vector<CGFloat> CopyLocations(const CGFloat* locations, size_t count) {
+        // locations can be null
+        RETURN_RESULT_IF(count == 0, {});
+        std::vector<CGFloat> result;
         if (locations) {
-            std::copy(locations, locations + count, std::back_inserter(_stopLocations));
-            return;
-        }
+            result = std::vector<CGFloat>(locations, locations + count);
+        } else {
+            result.reserve(count);
 
-        for (size_t i = 0; i < count; i++) {
-            _stopLocations[i] = (CGFloat)i / (CGFloat)(count - 1);
+            for (size_t i = 0; i < count; i++) {
+                result.push_back((CGFloat)i / (CGFloat)(count));
+            }
         }
+
+        return result;
     }
-};
 
-void __CGGradient::InitWithColorComponents(const CGFloat* components, const CGFloat* locations, size_t count) {
-    size_t componentCount = CGColorSpaceGetNumberOfComponents(_colorSpace) + 1;
-    _colorComponents.reserve(count * componentCount);
-    std::copy(components, components + (count * componentCount), std::back_inserter(_colorComponents));
-    SetLocations(locations, count);
-}
+    std::vector<CGFloat> CopyColorComponents(CGColorSpaceRef colorspace, CFArrayRef components) {
+        std::vector<CGFloat> result;
+        size_t componentCount = CGColorSpaceGetNumberOfComponents(colorspace) + 1;
+        CFIndex count = CFArrayGetCount(components);
+        result.reserve(count * componentCount);
+        static const CGFloat defaultColor[4] = { 0, 0, 0, 0 };
+
+        for (CFIndex i = 0; i < count; i++) {
+            CGColorRef curColor = (CGColorRef)CFArrayGetValueAtIndex(components, i);
```

Comment:
already added a comment regarding this, leaving it as a C cast.
We'll be doing too much dancing around to get the same.
Most of our code does a c-style cast for CFArratGetValue